src/simple_commander/game/game_controller.py

```diff
@@ -0,0 +1,190 @@
+        self.invaders_count = invaders_count
+        self.units = {}
+        self.random_type = self.get_unit_type()
+        self.set_invaders(self.invaders_count)
+
+    def __new__(cls, *args, **kwargs):
+        if not cls._instance:
+            cls._instance = super(GameController, cls).__new__(cls)
+        return cls._instance
+
+    def action(self, data):
+        for key in data:
+            action = getattr(self, key)
+            hero = self.units.get(data[key].get('id', ''), '')
+            if not action or not hero:
+                continue
+            if key == 'set_name':
+                action(hero, data[key].get('name', 'user'))
+            else:
+                action(hero)
+
+    def new_unit(self, unit_class, *args, **kwargs):
+        kwargs['controller'] = self
+        unit = unit_class(*args, **kwargs)
+        self.units[unit.id] = unit
+        unit.response('new')
+        unit.compute_new_coordinate(STEP_INTERVAL)
+        return unit
+
+    def drop_connection(self, socket):
+        socket = self.websockets[id(socket)]
+        self.remove_unit(socket['hero'])
+        del socket
+
+    @asyncio.coroutine
+    def notify_clients(self, data):
+        if self.websockets:
+            for key in self.websockets:
+                if not self.websockets[key]['socket']._closed:
+                    self.websockets[key]['socket'].send_str(json.dumps(data))
+
+    def new_hero(self):
+        pos_x = randint(0, self.game_field['width'])
+        pos_y = randint(0, self.game_field['height'])
+        angle = randint(0, 360)
+        hero_type = next(self.random_type)
+        hero = self.new_unit(Hero, x=pos_x, y=pos_y, angle=angle, obj_type=hero_type['type'],
+                             dimension=hero_type['dimension'])
+        return hero
+
+    def check_if_remove_units(self, units):
+        for unit in units:
+            if unit.is_dead:
+                self.remove_unit(unit.id)
+
+    def remove_unit(self, obj_id):
+        if self.units[obj_id]:
+            class_name = self.units[obj_id].__class__.__name__
+            self.units[obj_id].response('delete')
+            del self.units[obj_id]
+            if class_name == 'Invader':
+                self.set_invaders(1)
+
+    def start(self, socket, data, *args, **kwargs):
+        asyncio.async(self.run())
+        my_hero = self.new_hero()
+        self.websockets[id(socket)] = {'socket': socket, 'hero': my_hero.id}
+        name = data.get('name', 'user')
+        self.set_name(my_hero, name)
+        start_conditions = {'init': {
+            'hero_id': my_hero.id,
+            'game': self.game_field,
+            'units': self.get_units(),
+            'frequency': STEP_INTERVAL}}
+        socket.send_str(json.dumps(start_conditions))
+
+    def add_hits(self, bullet):
+        for unit in self.units:
+            if id(self.units[unit]) == bullet.unit_id and isinstance(self.units[unit], Hero):
```

Comment:
And, (maybe for future, but) I would use something like RulesBuilder or GameStrategy